### PR TITLE
[Lotto - Step 1] Submit mehrshahbaz's mission

### PR DESCRIPTION
**Pull request by MehrShahbaz**
This pull request adds a console-based lotto game. It covers the core game logic, input and output handling, and error management.

I would like feedback on the following questions:

- File structure
- Improvement on best practices of Kotlin
- Improvement on Design Pattern
- Improvement for test cases
- Improvement for functions designing

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,69 @@
...
+- [x] Calculate return rate
+
+### Lotto Handler - Wrapper Class
+- [x] Wrap the input/output handling and lotto game logic
+- [x] Reprompt on wrong input
+  - [x] Max 5 reprompts per input
+
+### Input View and Output View
+- [x] Create Input View
+- [x] Create Output View
```
Comment (Hyeon9mak) [inline README.md:34]: Looks great! 👍

Organizing business or feature requirements in documents like the README.md, will be helpful in the future whether it's Shahbaz revisiting the project later on, or a fellow developer joining the team and looking at the project for the first time. 

Hope you keep this habit going and keep refining it!

File: build.gradle.kts [hunk 2]
```diff
@@ -18,6 +18,7 @@ dependencies {
     testImplementation("org.junit.jupiter", "junit-jupiter", "5.10.2")
     testImplementation("org.assertj", "assertj-core", "3.25.3")
     testImplementation("io.kotest", "kotest-runner-junit5", "5.8.0")
+    testImplementation(kotlin("test"))
 }
```
Comment (Hyeon9mak) [inline build.gradle.kts:22]: I've always added the dependency with `testImplementation("org.jetbrains.kotlin:kotlin-test")`.
I didn't know about this approach!

This is much more Kotlin-like and cleaner 😄 👍

File: src/main/kotlin/lotto/LottoHandler.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            OutputView.displayWinnings(calculator.results)
+            OutputView.displayTotalWinningAmount(calculator.calculateTotalEarnings())
+            OutputView.displayReturnRate(returnRate)
+        } catch (err: IllegalArgumentException) {
+            println(err.message)
+        }
+    }
+
+    fun buyTickets(): LottoMachine {
+        repeat(MAX_ATTEMPT) {
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoHandler.kt]: this function is only used in `LottoHandler`, we can make it private.
IntelliJ highlighting this with yellow underlines.
It's always good to keep an eye on these IDE suggestions 😄

image

File: src/main/kotlin/lotto/LottoHandler.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            try {
+                val purchaseAmount = InputView.readPurchaseAmount()
+                val machine = LottoMachine(purchaseAmount)
+                OutputView.displayTickets(machine.tickets)
+                OutputView.displayChange(machine.showChange())
+                return machine
+            } catch (err: IllegalArgumentException) {
+                println(err.message)
+            }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoHandler.kt]: when switching from console screen to web browser,
it seems like both the View and LottoHandler need to change!

is it also the View's responsibility to decide whether to pass error information as strings or as images? 😄

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package lotto
+
+class LottoMachine(val purchaseAmount: Int, val tickets: MutableList<Lotto> = emptyList<Lotto>().toMutableList()) {
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt]:
```suggestion
class LottoMachine(
    val purchaseAmount: Int, 
    val tickets: MutableList<Lotto> = mutableListOf(),
) {
```
😉

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    private val numberList = (Lotto.Companion.MIN..Lotto.Companion.MAX)
+    private var change = 0
+    private var ticketCount = 0
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt]: As the name "companion object" suggests, we could naturally approach it this way!

```suggestion
    private val numberList = (Lotto.MIN..Lotto.MAX)
    private var change = 0
    private var ticketCount = 0
```

Moreover, if it's a fixed value that all object instances use identically...
how about placing it in the companion object area? 😉

```kotlin
companion object {
    private val LOTTO_NUMBER_RANGE = (Lotto.MIN..Lotto.MAX)
    // ...
}
```

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    init {
+        require(purchaseAmount in MIN..MAX) {
+            throw IllegalArgumentException()
+        }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt]: constructor based validation! 👍
the `require` already include throw exception 😄

image

and how about adding a message to help users quickly decide their next action?

```suggestion
    init {
        require(purchaseAmount in MIN..MAX) { "blah blah blah" }
```

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        generateTickets()
+    }
+
+    fun showChange() = change
+
+    fun generateTickets() {
+        repeat(ticketCount) {
+            tickets.add(Lotto(generateNumbers()))
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt]: using `repeat` very well!
and you can also use `List(n)` 😄 👍 
```kotlin
val lottos = List(ticketCount) { Lotto(generateNumbers()) }
tickets.addAll(lottos)
```

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    private fun generateNumbers(): List<Int> {
+        return numberList.shuffled().subList(0, Lotto.LOTTO_SIZE)
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt]: how about this? kotlin magic-
```suggestion
    private fun generateNumbers(): List<Int> {
        return numberList
            .shuffled()
            .take(Lotto.LOTTO_SIZE)
    }
```

File: src/main/kotlin/lotto/OutputView.kt [hunk 10]
```diff
@@ -0,0 +1,46 @@
+package lotto
+
+object OutputView {
+    fun displayTickets(tickets: List<Lotto>) {
+        println("You have purchased ${tickets.size} tickets.")
+        tickets.forEach { println(it.numbers.sorted().joinToString(prefix = "[", postfix = "]")) }
+        println()
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/OutputView.kt]: it doesn't have any properties.
it would be good to use the object class 👍 💯

File: src/main/kotlin/lotto/Rank.kt [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+            return when {
+                countOfMatch == FIRST.countOfMatch -> FIRST
+                countOfMatch == SECOND.countOfMatch && matchBonus -> SECOND
+                countOfMatch == THIRD.countOfMatch -> THIRD
+                countOfMatch == FOURTH.countOfMatch -> FOURTH
+                countOfMatch == FIFTH.countOfMatch -> FIFTH
+                else -> MISS
+            }
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Rank.kt:26]: in the future, if winningMoney changes or new rank are added, only need to look at the `Rank` enum class! 💯 👍

File: src/main/kotlin/lotto/WinningNumbers.kt [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+data class WinningNumbers(val winningNumbers: Lotto, val bonusNumber: Int) {
+    init {
+        require(bonusNumber in Lotto.MIN..Lotto.MAX) {
+            throw IllegalArgumentException()
+        }
+        require(!winningNumbers.numbers.contains(bonusNumber)) {
+            throw IllegalArgumentException()
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/WinningNumbers.kt]: check `require` & require kind message 🤣 

I think the `LottoNumber` and `BonusNumber` are the same thing.
How about wrapping both of them with the same VO (Value Object) and moving the validation there?

File: src/test/kotlin/lotto/CalculatorTest.kt [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+class CalculatorTest {
+    @Test
+    fun `Count number of matches between Lotto ticket and winning numbers`() {
+        val lotto = Lotto(listOf(3, 8, 27, 30, 35, 44))
+        val numbers = Lotto(listOf(3, 8, 27, 4, 5, 6))
+        val winningNumbers = WinningNumbers(numbers, 7)
+
+        val calculator = Calculator(mutableListOf<Lotto>(lotto), winningNumbers)
+        assertEquals(calculator.results.contains(Rank.FIFTH), true)
+    }
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/CalculatorTest.kt]: The three terms `lotto`, `numbers`, and `lottoTicket` are being used inconsistently.
How about unifying them into one consistent term?
This way, fellow developers won't get confused!

File: src/test/kotlin/lotto/LottoMachineTest.kt [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        val machine = LottoMachine(1234)
+        assertEquals(machine.showChange(), 234)
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = [1_000, 20_000])
+    fun `Generates correct number of tickets as a list`(amount: Int) {
+        val machine = LottoMachine(amount)
+        assertEquals(machine.tickets.size, amount / LottoMachine.TICKET_PRICE)
+    }
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/LottoMachineTest.kt]: Very detailed tests! 👍
How about also testing with boundary values?

```kotlin
    @ValueSource(ints = [-1000, 0, 999, 20_001, 21_000])
    fun `Purchase amount should be minimum 1_000 and (maximum 20_000 KRW)`(number: Int) {
```

File: src/main/kotlin/lotto/InputView.kt [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+    fun readWinningNumbers(): List<Int> {
+        println(WINNING_NUMBERS_MESSAGE)
+        val numbers =
+            readln().split(",")
+                .map {
+                    it.trim().toIntOrNull() ?: throw IllegalArgumentException(ERROR_MESSAGE)
+                }
+        return numbers
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/InputView.kt:18]: nice!
the View's responsibility to determine how to parse the string. 👍

**Review by Hyeon9mak — CHANGES_REQUESTED** [review]
Hello Shahbaz!

As you requested, I've focused my comments on 'Kotlin best practices' and 'test cases'.
For function design, design patterns, and file structure, since the right answer varies depending on each situation, I tried to respect Shahbaz's design as much as possible.

If I may make one suggestion, I believe the criteria for design should be business-driven, not technology-driven.
The design should be flexible enough to adapt to the business problems and characteristics that Shahbaz is trying to solve.

* What kind of design would be good for a business where Views change frequently but Domain changes are rare?
* Conversely, what kind of design would be good for a business where Views don't change but Domain changes are frequent?

With this kind of business context in mind, we can establish better design criteria.
During the HeroTechCourse mission, how about setting the standard as "a business where Views change frequently but Domain changes are rare"?
Or please establish your own criteria and explain the background to the reviewer!
I believe this will make for more enjoyable mission progress and code reviews.

I've left several comments, so please check them out.
Feel free to ask any questions you have!

File: README.md (hunk 1 above)
Comment (MehrShahbaz) [reply README.md:34]: Thank you for appreciation

I'll try to keep this habit up

File: src/main/kotlin/lotto/LottoHandler.kt (hunk 4 above)
Comment (MehrShahbaz) [reply src/main/kotlin/lotto/LottoHandler.kt]: I tried to move all the printing logic in the OutputVIew so that it is centralized

File: src/main/kotlin/lotto/InputView.kt [hunk 16]
```diff
@@ -0,0 +1,31 @@
+package lotto
+
+object InputView {
+    fun readPurchaseAmount(): Int {
+        println(PURCHASE_MESSAGE)
+        val amount = readln().trim().toIntOrNull() ?: throw IllegalArgumentException(ERROR_MESSAGE)
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/InputView.kt:6]: more user experience detail! 👍

File: src/main/kotlin/lotto/LottoHandler.kt [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        throw IllegalArgumentException(MAX_ATTEMPT_MESSAGE)
+    }
+
+    private fun processWinningNumbers(): Lotto {
+        repeat(MAX_ATTEMPT) {
+            try {
+                val winningNumbers = InputView.readWinningNumbers().map { it -> LottoNumber.from(it) }
+                return Lotto(winningNumbers)
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoHandler.kt:40]: how about this? 😉 

```suggestion
            try {
                val winningNumbers = InputView.readWinningNumbers().map(LottoNumber::from)
                return Lotto(winningNumbers)
```

File: src/main/kotlin/lotto/LottoHandler.kt [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        repeat(MAX_ATTEMPT) {
+            try {
+                val purchaseAmount = InputView.readPurchaseAmount()
+                val machine = LottoMachine(purchaseAmount)
+                OutputView.displayTickets(machine.tickets)
+                OutputView.displayChange(machine.showChange())
+                return machine
+            } catch (err: IllegalArgumentException) {
+                OutputView.displayError(err.message)
+            }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoHandler.kt:31]: 💯

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+        ticketCount = (purchaseAmount - change) / TICKET_PRICE
+        generateTickets()
+    }
+
+    fun showChange() = change
+
+    fun generateTickets() {
+        val lottoTickets = List(ticketCount) { Lotto(generateNumbers()) }
+        tickets.addAll(lottoTickets)
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt:22]: also it can change to private. only used in constructor.

File: src/main/kotlin/lotto/LottoMachine.kt [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+    private fun generateNumbers(): List<LottoNumber> =
+        numberList
+            .shuffled()
+            .take(Lotto.LOTTO_SIZE)
+
+    companion object {
+        private val numberList =
+            (LottoNumber.MINIMUM_NUMBER..LottoNumber.MAXIMUM_NUMBER)
+                .map { it -> LottoNumber.from(it) }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoMachine.kt:32]: Now the number range looks like static value.
Let's change the value name to uppercase! it'll be clearer for fellow developers.

Comment (Hyeon9mak) [reply src/main/kotlin/lotto/LottoMachine.kt:32]: and it can also skip the `it` keyword 😄

File: src/main/kotlin/lotto/Lotto.kt [hunk 21]
```diff
@@ -0,0 +1,16 @@
+package lotto
+
+class Lotto(val numbers: List<LottoNumber>) {
+    init {
+        require(numbers.size == LOTTO_SIZE)
+        require(numbers.distinct().size == LOTTO_SIZE)
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/Lotto.kt:7]: > how about adding a message to help users quickly decide their next action?

need more user experience detail!

```kotlin
require(numbers.size == LOTTO_SIZE) { "lotto numbers should be 6 numbers." }
```

File: src/main/kotlin/lotto/LottoNumber.kt [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+    companion object {
+        const val MINIMUM_NUMBER = 1
+        const val MAXIMUM_NUMBER = 45
+        private val NUMBERS: Map<Int, LottoNumber> =
+            (MINIMUM_NUMBER..MAXIMUM_NUMBER).associateWith(::LottoNumber)
+
+        fun from(value: Int): LottoNumber {
+            return NUMBERS[value] ?: throw IllegalArgumentException()
+        }
+    }
```
Comment (Hyeon9mak) [inline src/main/kotlin/lotto/LottoNumber.kt:13]: VO Caching 👍 👍 💯 
so now if `LottoNumber` data type change from `Int` to `Long`, just change `LottoNumber` class only! 😄

Comment (Hyeon9mak) [reply src/main/kotlin/lotto/LottoNumber.kt:13]: What are the differences between `class`, `value class`, `data class`?

File: src/test/kotlin/lotto/LottoNumberTest.kt [hunk 23]
```diff
@@ -0,0 +1,13 @@
...
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
+
+class LottoNumberTest {
+    @ParameterizedTest
+    @ValueSource(ints = [-5, 0, 51])
+    fun `Numbers should be between 0 and 50`(number: Int) {
+        assertThrows<IllegalArgumentException> { LottoNumber.from(number) }
+    }
+}
```
Comment (Hyeon9mak) [inline src/test/kotlin/lotto/LottoNumberTest.kt:13]: where is the success case? 🤔 
are you sure the `LottoNumber` should be between 0-50?

**Review by Hyeon9mak — APPROVED** [review]
Hi Shahbaz!
Awesome work on the step 1.

I have left some comments on a few simple improvements.
Let's fix them in the next step.

Feel free to ask any questions you have.
Have a nice weekend!
